content/lectures/lec02.md

```diff
@@ -6,6 +6,80 @@ assignments: ["hw0"]
 draft: false
 ---
 
+# Lecture Two: Docker
+
+## Portability
+
+Recall all of the issues that we've run into trying transfer code that we developed on our local machine to a remote host in the cloud. We want to be able to take our application code, and seamlessly run it on another machine. Along with Python dependencies, we needed to make sure that system resources, including Python and Poetry themselves, were also installed as expected. How can we avoid all of the dependency and versioning issues we ran into when we manually deployed to EC2?
+
+We were able to use Poetry and get all of the *reproducibility* benefits when we deployed to EC2, but just because a build is reproducible does *not* mean that it's **portable**.
+
+The intuition here is that we essentially want to _send a copy of our local computer_ to whoever is going to be running our code.
+
+## Containers
+
+Containers allow us to package minimal operating systems as "images" which includes everything our program needs to run. Once we've created the image, we can send it to anyone, and they will be able to build it on their machine to run our code. If you've used Virtual Machines in previous classes, you can think of container images as analogous to VM images, but with much less overhead.
```

Comment:
In the second sentence, I think "build" is the wrong word since in general an image is already built from a Dockerfile. Maybe just have it reference running the code?

Reply:
Oh, just read the next paragraph. Not really sure of how to make this clearer without destroying the abstraction and talking specifically about Docker

Reply:
see my comments on this part